content/blog/2019/nov/grpc-a-powerful-way-to-improve-golang-apis/index.md

```diff
@@ -0,0 +1,286 @@
+
+JSON is the most preferred way to return data from REST APIs. It is an excellent choice for public services used by front-end applications and third-parties. Due to its overhead on transporting and decoding, it is not well optimised for internal services communication.
+
+Even with these issues, it doesn't mean REST is a bad choice. It is by far the simplest way to implement an API, but there are known pitfalls due to its simplicity. Technologies such as GraphQL, Thrift and GRPC come to fill these gaps.
+
+# Enters GRPC and Protocol Buffers
+
+GRPC is a modern Remote Procedure Call (RPC) framework built on top-off HTTP 2.0, using Protocol Buffers as its interface definition language (IDL). These help it to be a low latency, high performance and scalable option to REST.
+
+The strict use of HTTP 2.0 enables developers to embrace some of its features, mainly streaming (client-side and server-side). For example:
+
+- A client can send a bulk of messages through a stream, instead of sending one request each or one with all of them
+- A client can send a request and receive a stream of responses, instead of waiting for the server to finish the processing and send a huge payload
+
+Protocol Buffers are a simpler and optimised way to define and serialise structured data. It can be used not only with GRPC but for other use cases, such as event modelling and data storage.
+
+Think Swagger (REST) or WSDL (SOAP), but lightweight due to the fact the payloads are binary, allowing faster serialisation and small footprint. Being generated by tools such as `protoc`, it brings some benefits such as type-safety in compile-time and auto-completion in IDEs.
+
+In the snippet below, a GRPC service is defined using Protobuf. Generating a client and server requires a simple `protoc` call. Easy right?
+
+```protobuf
+syntax = "proto3";
+
+package api;
+
+// Defines where your go package will be placed after compiled
+option go_package = "api";
+
+service Identity {
+  rpc GetUser (GetUserRequest) returns (GetUserResponse) {}
+  rpc GetUsers (GetUsersRequest) returns (GetUsersResponse) {}
+}
+
+message User {
+  string user_id = 1;
+  string name = 2;
+  bool active = 3;
+}
+
+message GetUserRequest {
+  string user_id = 1;
+}
+
+message GetUserResponse {
+  User user = 1;
+}
+
+message GetUsersRequest {}
+
+message GetUsersResponse {
+  // Defines an array of users
+  repeated User users = 1;
+}
+```
+
+In any protobuf file, each field has a number associated with it. These are unique identifiers in the binarised message and should not be changed. If one of these gets deprecated, they can't be re-used. It helps on avoiding breaking changes, allowing a progressive model evolution. Developers don't need to worry about fields been re-used, clashing with previous payloads, for example.
```

Comment:
These uniquely identify a position of a field in the binary message, so they shouldn’t be changed. ... fields being reused and clashing with previous versions of the message